### PR TITLE
Kademlia EachBin bug fix and unit tests

### PR DESCRIPTION
**Pull request by holisticode**
This PR adds a unit test for the `kademlia.EachBin` function.

This function is responsible for the correct registration of syncing subscriptions for the `stream` protocol.

Apart from fixing a couple of bugs, most importantly this PR introduces a unit test. 
https://github.com/ethersphere/go-ethereum/issues/1071 is an issue to refactor the `EachBin` function; with a working unit test that refactor gets a test-driven approach.

Replaces https://github.com/ethereum/go-ethereum/pull/18284 (which was for discussion)

### CODE REVIEW

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -367,12 +371,15 @@ func (k *Kademlia) EachBin(base []byte, pof pot.Pof, o int, eachBinFunc func(con
 	kadDepth := depthForPot(k.conns, k.MinProxBinSize, k.base)
 
 	k.conns.EachBin(base, pof, o, func(po, size int, f func(func(val pot.Val, i int) bool) bool) bool {
-		if startPo > 0 && endPo != k.MaxProxDisplay {
-			startPo = endPo + 1
-		}
+		//if the peer's bin is smaller than the kademlia depth,
```
Comment (zelig) [inline swarm/network/kademlia.go]: but now this PR really needs to wait for the new depth calculation

Comment (holisticode) [reply swarm/network/kademlia.go]: agreed

File: swarm/network/kademlia.go [hunk 2]
```diff
@@ -358,6 +358,10 @@ func (k *Kademlia) Off(p *Peer) {
 	}
 }
 
+//EachBin iterates over each connection of a kademlia table, per bin.
+//So `po` will be representing the bin value (0..255), while `val` represents an
+//existing connection inside that bin.
+//For any such connection found, the parameter func `eachBinFunc` will be executed
```
Comment (nolash) [inline swarm/network/kademlia.go]: I've also made a comment for this. Which one do we keep?

https://github.com/ethersphere/go-ethereum/pull/1069/files#diff-edfb4786e058d49a6335560cb9f86d00R362

File: swarm/network/kademlia.go (hunk 1 above)
Comment (nolash) [reply swarm/network/kademlia.go]: Please do _not_ use "smaller". Please review the terminology introduced here:

https://github.com/ethersphere/go-ethereum/pull/1051

If you don't agree with the terms, let's discuss and settle on something everyone can agree with.

File: swarm/network/kademlia_test.go [hunk 3]
```diff
@@ -731,3 +731,155 @@ func newTestDiscoveryPeer(addr pot.Address, kad *Kademlia) *Peer {
...
+009  0                              |  0
+010  0                              |  0
+011  0                              |  0
+012  0                              |  0
+013  0                              |  0
+014  0                              |  0
+015  0                              |  0
+=========================================================================
+*/
+func TestEachBin(t *testing.T) {
```
Comment (nolash) [inline swarm/network/kademlia_test.go]: A unit test should not be opinionated about the context which it's used by consumers. I find the terms "subscription" etc wholly out of place here. Keep it minimal to the exact expectations of what this function should provide and nothing more.

**Comment by holisticode** [comment]
has been obsoleted by https://github.com/ethereum/go-ethereum/pull/18355

**Pull request closed without merging**